src/analyses/startStateAnalysis.ml

```diff
@@ -0,0 +1,56 @@
+(** Remembers the abstract address value of each parameter at the beginning of each function by adding a ghost variable for each parameter.
+    Used by the c2po analysis. *)
+
+open GoblintCil
+open Batteries
+open Analyses
+open DuplicateVars.Var
+
+(**First all parameters (=formals) of the function are duplicated (by using DuplicateVars),
+    then we remember the value of each local variable at the beginning of the function
+  in this new duplicated variable. *)
+module Spec : Analyses.MCPSpec =
+struct
+  let name () = "startState"
+  module AD = ValueDomain.AD
+  module D = MapDomain.MapBot (Basetype.Variables) (AD)
+  module C = D
+
+  include Analyses.IdentitySpec
+
+  let ask_may_point_to (ask: Queries.ask) exp =
+    match ask.f (MayPointTo exp) with
+    | exception (IntDomain.ArithmeticOnIntegerBot _) -> AD.top()
+    | res -> res
+
+  let get_value (ask: Queries.ask) exp = ask_may_point_to ask exp
+
+  (** If e is a known variable (=one of the duplicated variables), then it returns the value for this variable.
+      If e is an unknown variable or an expression that is not simply a variable, then it returns top. *)
+  let eval (ask: Queries.ask) (d: D.t) (exp: exp): AD.t = match exp with
+    | Lval (Var x, NoOffset) -> begin match D.find_opt x d with
+        | Some v -> if M.tracing then M.trace "c2po-tainted" "QUERY %a : res = %a\n" d_exp exp AD.pretty v;v
+        | None -> AD.top()
+      end
+    | _ -> AD.top ()
+
+  let startcontext () = D.empty ()
+  let startstate v = D.bot ()
+  let exitstate = startstate
+
+  let query ctx (type a) (q: a Queries.t): a Queries.result =
+    let open Queries in
+    match q with
+    | MayPointTo e -> eval (ask_of_ctx ctx) ctx.local e
```

Comment:
How can this answer `MayPointTo` queries?

Reply:
This seems to only answer `MayPointTo` queries about ghost variables that are meant to capture information about the start state of the function.

However, it would be better to introduce a new type of query specifically for those ghost variables, something like `StartStateMayPointTo` or `GhostMayPointTo`

Reply:
The C-2PO analysis poses queries about expressions that may involve a ghost or a  non-ghost variable using `MayPointTo`, inside `CongurenceClosure.MayBeEqual`. Introducing a separation between queries for ghost variables and non-ghost variables would require to visit the expression that is passed to the query to evaluate whether it contains a ghost variable; and depending on the outcome dispatch a different query. 
It seems that this makes things more convulted.